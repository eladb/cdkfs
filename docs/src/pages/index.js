import React from 'react';
import clsx from 'clsx';
import Link from '@docusaurus/Link';
import useDocusaurusContext from '@docusaurus/useDocusaurusContext';
import Layout from '@theme/Layout';

import styles from './index.module.css';

function HomepageHeader() {
  const {siteConfig} = useDocusaurusContext();
  return (
    <header className={clsx('hero hero--primary', styles.heroBanner)}>
      <div className="container">
        <h1 className="hero__title">{siteConfig.title}</h1>
        <p className="hero__subtitle">{siteConfig.tagline}</p>
        <iframe width="560" height="315" src="https://www.youtube.com/embed/SOWMPzXtTCw" title="YouTube video player"
                frameBorder="0"
                allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share"
                allowFullScreen></iframe>
      </div>
    </header>
  );
}

export default function Home() {
  const {siteConfig} = useDocusaurusContext();
  return (
    <Layout
      title={`Hello from ${siteConfig.title}`}
      >
      <HomepageHeader />
      <main>
      </main>
    </Layout>
  );
}
