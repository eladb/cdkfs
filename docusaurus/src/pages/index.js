import React from 'react';
import useDocusaurusContext from '@docusaurus/useDocusaurusContext';
import Layout from '@theme/Layout';
import Image1 from '../../static/img/backgrounds/1.png';
import Image2 from '../../static/img/backgrounds/2.png';
import TypeScriptLogo from '../../static/img/languages/ts.png';
import GoLogo from '../../static/img/languages/go.png';
import JavaLogo from '../../static/img/languages/java.png';
import PythonLogo from '../../static/img/languages/python.png';

export default function Home() {
    const {siteConfig} = useDocusaurusContext();
    return (
        <Layout
            title={siteConfig.title}
        >
            <main>
                {/* SECTION 1 */}
                <section className="bg-slate-800 text-white py-16 lg:py-36 relative overflow-hidden">
                    <div className="container mx-auto text-center">
                        <img src={siteConfig.themeConfig.image}
                             alt="The projen logo containing three boxes stacked in a pyramid shape with the bottom right box having the letters 'pj' in it"
                             className="mx-auto mb-8 lg:w-36 relative z-10"/>
                        <h1 className="text-6xl font-bold mb-8 relative z-10">{siteConfig.title}</h1>
                        <div className="container mx-auto text-center relative z-10">
                            <h2 className="text-2xl lg:text-5xl font-medium text-white mb-16 lg:mb-24">
                                Rapidly build modern applications
                                <span className="lg:block mt-3"> with advanced configuration management</span>
                            </h2>
                            <a href={`${siteConfig.baseUrl}docs/getting-started/`}
                               className="rounded text-white bg-blue-500 hover:text-white py-3 px-8 inline-block lg:mb-36">Get
                                Started</a>
                        </div>
                        <img src={Image1}
                             alt="An illustration containing various lines, shapes in blues, reds and whites making a larger cube centered inside of a circular pattern."
                             className="mx-auto mb-8 opacity-25 absolute -left-48 top-24 lg:top-48 z-0"/>

                    </div>
                </section>

                <hr className="m-0 bg-slate-600"/>

                {/* SECTION 2 */}
                <section className="dark:bg-slate-800 pt-16 pb-36 lg:py-36">
                    <div className="p-8 lg:p-0 lg:container lg:mx-auto text-center relative z-10">
                        <h2 className="text-2xl lg:text-5xl font-medium dark:text-white mb-16 lg:mb-24">
                            Build and manage your
                            <span className="lg:block mt-3"> apps with confidence</span>
                        </h2>

                        <div className="lg:grid lg:grid-cols-4">
                            <div className="lg:col-start-2 lg:col-span-2">
                                <p className="text-xl lg:text-2xl text-center">
                                    Synthesize project configuration files from a well-typed definitions, written in
                                    {' '}<a href="https://github.com/aws/jsii" target="_blank"
                                            className="text-blue-500 dark:text-blue-300 underline">jsii</a>{' '}
                                    compatible languages reducing project configuration management fatigue, increasing
                                    productivity and confidence.
                                </p>
                            </div>
                        </div>
                    </div>
                </section>

                <hr className="m-0 bg-slate-600"/>

                {/* SECTION 3 */}
                <section className="dark:bg-slate-800 pt-16 pb-36 lg:py-36">
                    <div className="p-8 lg:p-0 lg:container lg:mx-auto text-center relative z-10">
                        <h2 className="text-2xl lg:text-5xl font-medium dark:text-white mb-16 lg:mb-24">
                            Available for these
                            <span className="lg:block mt-3"> popular languages</span>
                        </h2>

                        <ul className="lg:grid lg:grid-cols-5 lg:gap-4 lg:items-center p-0">
                            <li className="list-none mb-12 lg:mb-0 text-center">
                                <img src={TypeScriptLogo} alt="TypeScript" className="h-48"/>
                            </li>
                            <li className="list-none mb-12 lg:mb-0 text-center">
                                <img src={GoLogo} alt="Go" className="h-48 w-auto"/>
                            </li>
                            <li className="list-none mb-12 lg:mb-0 text-center">
                                <img src={JavaLogo} alt="Java" className="h-48 w-auto"/>
                            </li>
                            <li className="list-none mb-12 lg:mb-0 text-center">
                                <img src={PythonLogo} alt="Python" className="rounded"/>
                            </li>
                        </ul>
                    </div>
                </section>

                <hr className="m-0 bg-slate-600"/>

                {/* SECTION 4 */}
                <section className="dark:bg-slate-800 pt-16 pb-36 lg:py-36 relative overflow-hidden">
                    <div className="p-8 lg:p-0 lg:container lg:mx-auto text-center relative z-10">
                        <h2 className="text-2xl lg:text-5xl font-medium dark:text-white mb-16 lg:mb-24">
                            Extensible, customizable
                            <span className="lg:block mt-3"> and adaptable</span>
                        </h2>

                        <div className="lg:grid lg:grid-cols-4">
                            <div className="lg:col-start-2 lg:col-span-2">
                                <p className="text-xl lg:text-2xl text-center">
                                    Tasks allow you to define commands backed by shell scripts. You can use tasks to implement custom
                                    workflows accessible through the projen CLI. Focus more on business requirements,
                                    less on managing project configuration.
                                </p>
                            </div>
                        </div>
                    </div>
                    <img src={Image2}
                         alt="A vector illustration of a cube made from red, yellow and blue puzzle pieces."
                         className="mx-auto mb-8 opacity-[.10] lg:opacity-25 absolute -right-48 top-24 lg:top-48 z-0"/>
                </section>

                <hr className="m-0 bg-slate-600"/>

                {/* SECTION 5 */}
                <section className="dark:bg-slate-800 pt-16 pb-36 lg:py-36 relative overflow-hidden">
                    <div className="p-8 lg:p-0 lg:container lg:mx-auto text-center relative z-10">
                        <h2 className="text-2xl lg:text-5xl font-medium dark:text-white mb-16 lg:mb-24">
                            Built by developers,
                            <span className="lg:block mt-3"> for developers</span>
                        </h2>

                        <div className="lg:grid lg:grid-cols-4 mb-8">
                            <div className="lg:col-start-2 lg:col-span-2">
                                <p className="text-xl lg:text-2xl text-center">
                                    Youâ€™re not alone. Developers from all over the world manage their complex configurations using
                                    projen. <a href={`${siteConfig.baseUrl}docs/getting-started/`} className="text-blue-500 dark:text-blue-300 underline">Get started</a> today
                                    and get these benefits:
                                </p>
                            </div>
                        </div>

                        <ul className="md:grid md:grid-cols-12 md:gap-4 pl-0">
                            <li className="list-none md:col-span-6 rounded border-solid dark:border-slate-600 dark:bg-slate-700 p-6 mb-8 lg:mb-0">
                                Standard npm scripts like compile, build, test, package and more.
                            </li>
                            <li className="list-none md:col-span-6 rounded border-solid dark:border-slate-600 dark:bg-slate-700 p-6 mb-8 lg:mb-0">
                                jsii: compile, package, api compatibility checks, API.md
                            </li>
                            <li className="list-none md:col-span-4 rounded border-solid dark:border-slate-600 dark:bg-slate-700 p-6 mb-8 lg:mb-0">
                                Fully synthesize package.json
                            </li>
                            <li className="list-none md:col-span-4 rounded border-solid dark:border-slate-600 dark:bg-slate-700 p-6 mb-8 lg:mb-0">
                                Automated PR builds
                            </li>
                            <li className="list-none md:col-span-4 rounded border-solid dark:border-slate-600 dark:bg-slate-700 p-6 mb-8 lg:mb-0">
                                Automated dependency upgrades
                            </li>
                            <li className="list-none col-span-6 rounded border-solid dark:border-slate-600 dark:bg-slate-700 p-6 mb-8 lg:mb-0">
                                Bump & release scripts with CHANGELOG generation based on conventional commits
                            </li>
                            <li className="list-none col-span-6 rounded border-solid dark:border-slate-600 dark:bg-slate-700 p-6 mb-8 lg:mb-0">
                                Node "engines" support with coupling to CI build environment and @types/node
                            </li>
                        </ul>
                    </div>
                </section>
            </main>
        </Layout>
    );
}
