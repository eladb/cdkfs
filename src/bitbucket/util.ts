

export function reduceRenderObject( obj:Record<string, any> ) {

  const result = Object.keys(obj).reduce( ( inp:Record<string, any>, key:string) => {

    if (obj[ key ] !== undefined) inp[ key ] = obj[ key ];

    return inp;
  }, {} );

  if ( Object.keys( result).length === 0 ) return undefined;

  return result;
}


export function reduceRenderArray( arr:any[]|undefined ) {

  if ( arr === undefined ) return arr;
  else if (arr.length === 0) return undefined;
  return arr;
}