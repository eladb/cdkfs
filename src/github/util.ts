export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

/**
 * Bash if equivalent execution of scripts that is compatible with shx.
 * Execution will fail, if the when or then command fails.
 *
 * @param when The condition to evaluate. Usually a `test` command. This command may be executed twice. Make sure it is non mutating.
 * @param then The script to run if the condition is true.
 * @param otherwise The script to run if the condition is false.
 */
export function shxIf(when: string, then: string, otherwise?: string): string {
  // Connect the condition and the 'then' command with a logical AND.
  // If the condition is true, this will execute the 'then' command.
  // This case is equivalent to TRUE && then()
  // If the condition is false, this will exit with a non-zero status code
  // (indicating an error) without executing the 'then' command.
  // This case is equivalent to FALSE && then()
  const if_clause = `${when} && (${then})`;

  // For the else clause, we need to check the 'when' condition again,
  // because the if clause can also return FALSE if the 'then' command fails.
  // However in this situation we don't want to execute the else clause but report an error.
  const else_clause = `${when} || (${otherwise})`;

  // Try the if_clause first. If it fails try the else clause.
  if (otherwise) {
    return `(${if_clause}) || (${else_clause})`;
  }

  // Just run the if clause
  return if_clause;
}
