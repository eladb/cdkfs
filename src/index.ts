export * from './file';
export * from './github-workflow';
export * from './ignore-file';
export * from './jsii-project';
export * from './json';
export * from './license';
export * from './node-project';
export * from './project';
export * from './component';
export * from './semver';
export * from './version';
export * from './eslint';
export * from './jest';
export * from './typescript';
export * from './mergify';
export * from './dependabot';