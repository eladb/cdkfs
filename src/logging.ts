import * as chalk from 'chalk';

let enabled = true;

function log(color: chalk.ChalkFunction, ...text: any[]) {
  if (!enabled) { return; }
  console.error(`ðŸ¤– ${color(...text)}`);
}

/**
 * @deprecated
 */
export function debug(...text: any[]) {
  if (process.env.DEBUG) {
    log(chalk.gray, ...text);
  }
}

/**
 * @deprecated
 */
export function verbose(...text: any[]) {
  log(chalk.white, ...text);
}

/**
 * @deprecated
 */
export function info(...text: any[]) {
  log(chalk.cyan, ...text);
}

/**
 * @deprecated
 */
export function error(...text: any[]) {
  log(chalk.red, ...text);
}

/**
 * @deprecated
 */
export function warn(...text: any[]) {
  log(chalk.yellow, ...text);
}

/**
 * @deprecated
 */
export function disable() {
  enabled = false;
}
