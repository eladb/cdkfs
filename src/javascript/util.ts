import { existsSync, readFileSync } from "fs";
import { basename, dirname, extname, join, sep, resolve } from "path";
import * as semver from "semver";
import { findUp } from "../util";

/**
 * Basic interface for `package.json`.
 */
interface PackageManifest {
  [key: string]: any;

  /**
   * Package name.
   */
  name: string;
  /**
   * Package version.
   */
  version: string;
}

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>[^\/.]+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve location of the given `moduleId`.
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModule(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  try {
    return require.resolve(moduleId, options);
  } catch {
    return undefined;
  }
}

/**
 * Attempt to resolve a module's manifest (package.json) path via `require.resolve` lookup.
 *
 * @remarks
 * If the target package has `exports` that differ from the default
 * (i.e, it defines the `exports` field in its manifest) and does not
 * explicitly include an entry for `package.json`, this strategy will fail.
 * See {@link tryResolveManifestPathFromDefaultExport} as an alternative.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModuleManifestPath(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  // cannot just `require('dependency/package.json')` here because
  // `options.paths` may not overlap with this node proc's resolution paths.
  const manifestId = `${moduleId}/package.json`;
  return tryResolveModule(manifestId, options);
}

/**
 * Attempt to resolve a module's manifest (package.json) path by looking for the nearest
 * `package.json` file that is an ancestor to the module's default export location.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveManifestPathFromDefaultExport(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const defaultExportPath = tryResolveModule(moduleId, options);
  if (!defaultExportPath) {
    return undefined;
  }
  const moduleDir = findUp("package.json", defaultExportPath);
  if (!moduleDir) {
    return undefined;
  }
  return join(moduleDir, "package.json");
}

/**
 * Attempt to resolve a module's manifest (package.json) path by checking for its existence under `node_modules` relative to `basePath`.
 *
 * @remarks
 * This strategy can be helpful in the scenario that a module defines
 * custom exports without `package.json` and no default export (i.e, some type definition packages).
 *
 * @param moduleId Module ID to lookup.
 * @param basePath Root path to search from.
 */
export function tryResolveManifestPathFromPath(
  moduleId: string,
  basePath: string
) {
  const base = basePath.includes("node_modules")
    ? basePath
    : join(basePath, "node_modules");
  const filePath = resolve(base, ...moduleId.split("/"), "package.json");
  if (existsSync(filePath)) {
    return filePath;
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) path by searching for it in the optionally provided paths array
 * as well as the current node processes' default resolution paths.
 * @param moduleId Module ID to search for.
 * @param options Search options.
 */
export function tryResolveManifestPathFromSearch(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const searchPaths = [
    ...(options?.paths ?? []),
    ...(require.resolve.paths(moduleId) ?? []),
  ];
  for (const path of searchPaths) {
    const result = tryResolveManifestPathFromPath(moduleId, path);
    // early return on first result.
    if (result) {
      return result;
    }
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) using multiple strategies.
 * @param moduleId Module to resolve manifest path for.
 * @param options Resolution options.
 */
export function tryResolveModuleManifest(
  moduleId: string,
  options?: { paths: string[] }
): PackageManifest | undefined {
  const strategies = [
    tryResolveModuleManifestPath,
    tryResolveManifestPathFromDefaultExport,
    tryResolveManifestPathFromSearch,
  ];
  for (const strategy of strategies) {
    const result = strategy(moduleId, options);
    // early return on first result.
    if (result) {
      try {
        const manifest = JSON.parse(
          readFileSync(result, "utf8")
        ) as PackageManifest;
        // verify name matches target module.
        if (manifest.name === moduleId) {
          return manifest;
        }
      } catch {
        // continue to next strategy.
      }
    }
  }
  return undefined;
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifest = tryResolveModuleManifest(dependencyName, options);
  if (!manifest) {
    return undefined;
  }
  return manifest?.version;
}

/**
 * Whether the given requestedRange *probably* leads to the installation of a version that matches checkRange
 *
 * We assume that NPM always installs the most recent version of a package that
 * is allowed by the requestedRange.
 */
export function installedVersionProbablyMatches(
  requestedRange: string,
  checkRange: string
): boolean {
  const options = {
    includePrerelease: true,
    loose: true,
  };

  // No questions asked: always true
  if (semver.subset(requestedRange, checkRange, options)) {
    return true;
  }

  // Also no questions asked: always false
  if (!semver.intersects(requestedRange, checkRange, options)) {
    return false;
  }

  // Now we're in tricky territory. We intersect, but aren't a full subset.
  // We are in one of the following 2 situations, which we will tie-break by
  // assuming NPM will install the most recent matching version in 'requested'.
  //
  // requested  | check    | result
  // -----------|----------|-----------
  //   >= 2     |  >= 3    | probably true (chance of FP)
  //   <= 2     |  <= 1    | probably false (change of FN)
  //
  // `semver` doesn't make it easy to distinguish these cases (we can't request
  // the `maxVersion` that satisfies a range). Instead what we do is
  // get the `minVersion` of each range, and if they compare equal we assume
  // we're in the bottom case with `<=` checks, and return `false`.

  return !semver.eq(
    semver.minVersion(requestedRange, options) ?? "1.2.3",
    semver.minVersion(checkRange, options) ?? "1.2.3"
  );
}
