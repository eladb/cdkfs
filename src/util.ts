import * as child_process from 'child_process';
import * as path from 'path';
import * as fs from 'fs-extra';
import * as logging from './logging';
// eslint-disable-next-line @typescript-eslint/no-require-imports
const decamelize = require('decamelize');

export function exec(command: string, options?: child_process.ExecSyncOptions) {
  logging.verbose(command);
  return child_process.execSync(command, {
    stdio: ['inherit', process.stderr, 'pipe'],
    ...options,
  });
}

/**
 * Executes `command` and returns its value or undefined if the command failed.
 */
export function execOrUndefined(command: string, options?: child_process.ExecSyncOptions): string | undefined {
  try {
    const value = child_process.execSync(command, { stdio: ['inherit', 'pipe', 'ignore'], ...options }).toString('utf-8').trim();
    if (!value) { return undefined; } // an empty string is the same as undefined
    return value;
  } catch {
    return undefined;
  }
}

export interface WriteFileOptions {
  /**
   * Whether the generated file should be marked as executable.
   *
   * @default false
   */
  executable?: boolean;

  /**
   * Whether the generated file should be readonly.
   *
   * @default false
   */
  readonly?: boolean;
}

export function getFilePermissions(options: WriteFileOptions): string {
  const readonly = options.readonly ?? false;
  const executable = options.executable ?? false;
  if (readonly && executable) {
    return '500';
  } else if (readonly) {
    return '400';
  } else if (executable) {
    return '755';
  } else {
    return '644';
  }
}

export function writeFile(filePath: string, data: any, options: WriteFileOptions = {}) {
  if (fs.existsSync(filePath)) {
    fs.chmodSync(filePath, '600');
  }

  fs.mkdirpSync(path.dirname(filePath));
  fs.writeFileSync(filePath, data);

  fs.chmodSync(filePath, getFilePermissions(options));
}

/**
 * Decamelizes the keys of an object structure, recursing through child objects and arrays.
 * @experimental
 */
export interface DecamelizeRecursivelyOptions {
  /**
   * Max depth to recurse before erroring.
   * @default 10
   */
  maxDepth?: number;

  /**
   * Returns true when a key should be decamelized
   * @default - all keys are decamelized
   */
  shouldDecamelize?: (path: string[], value: any) => boolean;

  /**
   * Separator for decamelizing.
   * @default "_"
   */
  separator?: string;

  /**
   * Current path.
   * @internal
   */
  path?: string[];
}

export function decamelizeKeysRecursively(input: any, opt?: DecamelizeRecursivelyOptions): any {
  const shouldAlwaysDecamelize = () => true;
  const shouldDecamelize = opt?.shouldDecamelize ?? shouldAlwaysDecamelize;
  const separator = opt?.separator ?? '_';
  const path_ = opt?.path ?? [];
  const maxDepth = opt?.maxDepth ?? 10;

  if (path_.length > maxDepth) {
    throw new Error('Decamelled too deeply - check that the input has no circular references');
  }

  if (Array.isArray(input)) {
    return input.map((k, i) => decamelizeKeysRecursively(k, {
      ...opt,
      path: [...path_, i.toString()],
    }));
  }

  if (typeof input === 'object' && input !== null) {
    const mappedObject: Record<string, any> = {};
    for (const [key, value] of Object.entries(input)) {
      const transformedKey = shouldDecamelize([...path_, key], value)
        ? decamelize(key, separator)
        : key;

      mappedObject[transformedKey] = decamelizeKeysRecursively(value, {
        ...opt,
        path: [...path_, key],
      });
    }

    return mappedObject;
  }

  return input;
}

/**
 * Returns false if value is unset or a falsey value, and true otherwise.
 * @param value an environment variable
 */
export function isTruthy(value: string | undefined): boolean {
  return !(value === undefined || ['null', 'undefined', '0', 'false', ''].includes(value.toLocaleLowerCase()));
}

/**
 * Type of a map mapping strings to some arbitrary type
 */
export type Obj<T> = { [key: string]: T };

/**
 * Return whether the given value is an object
 *
 * Even though arrays and instances of classes technically are objects, we
 * usually want to treat them differently, so we return false in those cases.
 */
export function isObject(x: any): x is Obj<any> {
  return x !== null && typeof x === 'object' && !Array.isArray(x)
    && x.constructor.name === 'Object';
}

/**
 * Recursively merge objects together
 *
 * The leftmost object is mutated and returned. Arrays are not merged
 * but overwritten just like scalars.
 *
 * If an object is merged into a non-object, the non-object is lost.
 *
 * `undefined`s will cause a value to be deleted if destructive is enabled.
 */
export function deepMerge(objects: Array<Obj<any> | undefined>, destructive: boolean = false) {
  function mergeOne(target: Obj<any>, source: Obj<any>) {
    for (const key of Object.keys(source)) {
      const value = source[key];

      if (isObject(value)) {
        // if the value at the target is not an object, override it with an
        // object so we can continue the recursion
        if (!isObject(target[key])) {
          target[key] = value;
        }
        mergeOne(target[key], value);

        // if the result of the merge is an empty object, it's because the
        // eventual value we assigned is `undefined`, and there are no
        // sibling concrete values alongside, so we can delete this tree.
        const output = target[key];
        if (isObject(output) && Object.keys(output).length === 0 && destructive) {
          delete target[key];
        }
      } else if (value === undefined && destructive) {
        delete target[key];
      } else if (typeof value !== 'undefined') {
        target[key] = value;
      }
    }
  }

  const others = objects.filter(x => x != null) as Array<Obj<any>>;

  if (others.length === 0) { return {}; }
  const into = others.splice(0, 1)[0];

  others.forEach(other => mergeOne(into, other));
  return into;
}

/*
 * Deduplicate values in a list, returning a new array.
 * @param array list of values
 */
export function dedupArray<T>(array: T[]): T[] {
  return array.filter((val, idx) => array.indexOf(val) === idx);
}

/**
 * Returns a sorted version of `x` or `undefined` if it is an empty array or object.
 */
export function sorted<T>(x: T) {
  if (x == null) { return undefined; }
  if (Array.isArray(x)) {
    if (x.length === 0) { return undefined; }
    return (x as unknown[]).sort();
  } else if (typeof x === 'object') {
    if (Object.keys(x).length === 0) { return undefined; }
    const result: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(x).sort(([l], [r]) => l.localeCompare(r))) {
      result[key] = value;
    }
    return result as T;
  } else {
    return x;
  }
}

export function formatAsPythonModule(name: string) {
  return name.replace(/-/g, '_').replace(/\./g, '_');
}
