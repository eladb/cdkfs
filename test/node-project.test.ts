import { NodeProject } from '../src';
import * as logging from '../src/logging';
import { mkdtemp, synthSnapshot } from './util';

logging.disable();

test('license file is added by default', () => {
  // WHEN
  const project = new NodeProject({
    outdir: mkdtemp(),
    name: 'test-node-project',
    mergify: false,
    projenDevDependency: false,
  });

  // THEN
  expect(synthSnapshot(project).LICENSE).toContain('Apache License');
});

test('license file is not added if licensed is false', () => {
  // WHEN
  const project = new NodeProject({
    outdir: mkdtemp(),
    name: 'test-node-project',
    licensed: false,
    mergify: false,
    projenDevDependency: false,
  });

  // THEN
  const snapshot = synthSnapshot(project);
  expect(snapshot.LICENSE).toBeUndefined();
  expect(snapshot['.gitignore']).not.toContain('LICENSE');
  expect(snapshot['package.json'].license).toEqual('UNLICENSED');
});
