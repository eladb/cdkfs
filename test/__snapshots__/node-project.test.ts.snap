// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deps upgrade commit can be signed 1`] = `
Object {
  "if": "\${{ needs.upgrade.outputs.patch_created }}",
  "name": "Create Pull Request",
  "needs": "upgrade",
  "permissions": Object {
    "contents": "write",
    "pull-requests": "write",
  },
  "runs-on": "ubuntu-latest",
  "steps": Array [
    Object {
      "name": "Checkout",
      "uses": "actions/checkout@v2",
      "with": Object {
        "ref": "main",
        "token": "\${{ secrets.PROJEN_GITHUB_TOKEN }}",
      },
    },
    Object {
      "name": "Download patch",
      "uses": "actions/download-artifact@v2",
      "with": Object {
        "name": ".repo.patch",
        "path": "\${{ runner.temp }}",
      },
    },
    Object {
      "name": "Apply patch",
      "run": "[ -s \${{ runner.temp }}/.repo.patch ] && git apply \${{ runner.temp }}/.repo.patch || echo \\"Empty patch. Skipping.\\"",
    },
    Object {
      "name": "Set git identity",
      "run": "git config user.name \\"github-actions\\"
git config user.email \\"github-actions@github.com\\"",
    },
    Object {
      "id": "create-pr",
      "name": "Create Pull Request",
      "uses": "peter-evans/create-pull-request@v3",
      "with": Object {
        "author": "github-actions <github-actions@github.com>",
        "body": "Upgrades project dependencies. See details in [workflow run].

[Workflow Run]: https://github.com/\${{ github.repository }}/actions/runs/\${{ github.run_id }}

------

*Automatically created by projen via the \\"upgrade-main\\" workflow*",
        "branch": "github-actions/upgrade-main",
        "commit-message": "chore(deps): upgrade dependencies

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]: https://github.com/\${{ github.repository }}/actions/runs/\${{ github.run_id }}

------

*Automatically created by projen via the \\"upgrade-main\\" workflow*",
        "committer": "github-actions <github-actions@github.com>",
        "signoff": true,
        "title": "chore(deps): upgrade dependencies",
        "token": "\${{ secrets.PROJEN_GITHUB_TOKEN }}",
      },
    },
  ],
}
`;

exports[`mutableBuild will push changes to PR branches 1`] = `
Array [
  Object {
    "name": "Checkout",
    "uses": "actions/checkout@v2",
    "with": Object {
      "ref": "\${{ github.event.pull_request.head.ref }}",
      "repository": "\${{ github.event.pull_request.head.repo.full_name }}",
    },
  },
  Object {
    "name": "Install dependencies",
    "run": "yarn install --check-files",
  },
  Object {
    "name": "build",
    "run": "npx projen build",
  },
  Object {
    "id": "self_mutation",
    "name": "Find mutations",
    "run": "git add .
git diff --staged --patch --exit-code > .repo.patch || echo \\"::set-output name=self_mutation_happened::true\\"",
  },
  Object {
    "if": "steps.self_mutation.outputs.self_mutation_happened",
    "name": "Upload patch",
    "uses": "actions/upload-artifact@v2",
    "with": Object {
      "name": ".repo.patch",
      "path": ".repo.patch",
    },
  },
  Object {
    "if": "steps.self_mutation.outputs.self_mutation_happened",
    "name": "Fail build on mutation",
    "run": "echo \\"::error::Files were changed during build (see build log). If this was triggered from a fork, you will need to update your branch.\\"
cat .repo.patch
exit 1",
  },
]
`;
